### PR TITLE
[WIP] Revised SE-0107: UnsafeRawPointer API, initialization method names.

### PR DESCRIPTION
**Pull request by atrick**
Naming improvements suggested by Brent Royal-Gordon.

### CODE REVIEW

File: proposals/0107-unsaferawpointer.md [hunk 1]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
   func bindMemory<T>(to: T.Type, capacity: Int) -> UnsafeMutablePointer<T>
   func assumingMemoryBound<T>(to: T.Type) -> UnsafeMutablePointer<T>
 
-  func initialize<T>(_: T.Type, with: T, count: Int = 1)
+  func initialize<T>(_: T.Type, to: T, count: Int = 1)
     -> UnsafeMutablePointer<T>
-  func initialize<T>(toContiguous: T.Type, atIndex: Int, with: T)
+  func initialize<T>(contiguous: T.Type, at: Int, to: T)
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: This one looks really easy to misuse.  Is the `at` parameter (BTW, you **need** parameter names distinct from the argument labels!) measured in bytes or in `T`'s?

Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Both of the above don’t look like they read well:

```swift
address.initialize(String.self, to: “hello")
address.initialize(contiguous: String.self, at: position, to: “hello”)
```
In the first one, it makes `String.self` read as the direct object, but we are not initializing the type; we’re initializing memory.

`contiguous` is very confusing in the latter call because it implies multiple elements, yet this initializes only a single element.  What’s it supposed to mean?

File: proposals/0107-unsaferawpointer.md [hunk 2]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
-  func initializeBackward<T>(from: UnsafePointer<T>, count: Int)
+  func initialize<T>(from: UnsafePointer<T>, backwardToCount: Int)
     -> UnsafeMutablePointer<T>
 
   // The `move` APIs deinitialize the memory at `from`.
-  func moveInitialize<T>(from: UnsafePointer<T>, count: Int)
+  func moveInitialize<T>(from: UnsafePointer<T>, forwardToCount: Int)
     -> UnsafeMutablePointer<T>
-  func moveInitializeBackward<T>(from: UnsafePointer<T>, count: Int)
+  func moveInitialize<T>(from: UnsafePointer<T>, backwardToCount: Int)
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Didn’t we agree to dump separate forward/backward APIs?

File: proposals/0107-unsaferawpointer.md [hunk 3]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
     -> UnsafeMutablePointer<T>
 
-  func load<T>(_: T.Type) -> T
-  func load<T>(_: T.Type, atByteOffset: Int) -> T
-  func load<T>(fromContiguous: T.Type, atIndex: Int) -> T
+  func load<T>(as: T.Type) -> T
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: I don’t think we want a label here.  Isn’t it reasonable to say

```swift
address.load(String.self)
```
?

File: proposals/0107-unsaferawpointer.md [hunk 4]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
+  func load<T>(fromByteOffset: Int, as: T.Type) -> T
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: I like the old one better.  The change from `at` to `from` is nice, though.

File: proposals/0107-unsaferawpointer.md [hunk 5]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
+  func load<T>(fromContiguous: T.Type, at: Int) -> T
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: I don’t know what contiguous means in this context.  BTW, doc comments would help these APIs.  You shouldn’t really propose APIs without them.

File: proposals/0107-unsaferawpointer.md [hunk 6]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
 
   // storeRaw performs bytewise writes, but proper alignment for `T` is still
   // required.
   // T must be a trivial type.
-  func storeRaw<T>(_: T.Type, with: T)
-  func storeRaw<T>(toContiguous: T.Type, atIndex: Int, with: T)
-  func storeRaw<T>(contiguous: T.Type, from: UnsafeRawPointer, count: Int)
-  func storeRawBackward<T>(
-    contiguous: T.Type, from: UnsafeRawPointer, count: Int)
+  func storeRaw<T>(_: T, as: T.Type)
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Why are we putting “raw” in the name?  You’re already in a “raw” pointer.  Presumably all the operations are somehow raw.

File: proposals/0107-unsaferawpointer.md [hunk 7]
```diff
@@ -1587,12 +1589,14 @@ extension UnsafeMutablePointer<Pointee> {
   func deallocate(capacity: Int)
 
   // Naming conventions.
```
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: What does that comment mean?

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (brentdax) [inline proposals/0107-unsaferawpointer.md]: I read it as "initialize contiguous `String` at `position` to `"hello"`"—that is, assume the pointer points to contiguous `String`s, and initialize the `String` at `position` to `"hello"`.

If you don't think that reads well, I wonder if we can combine both the `at` and `count` variants into one call:

```swift
func initialize<T>(_: T.Type, at: Int = 0, count: Int = 1, to value: T) -> UnsafeMutablePointer<T>

address.initialize(String.self, to: "hello")
address.initialize(String.self, at: 3, to: "hello")
address.initialize(String.self, count: 3, to: "hello")
address.initialize(String.self, at: 3, count: 3, to: "hello") 
```

If the bare type doesn't read well, it could have an `as` label.

File: proposals/0107-unsaferawpointer.md (hunk 6 above)
Comment (brentdax) [inline proposals/0107-unsaferawpointer.md]: I'm not the one who chose this name, but I read it as a reminder that this API won't do reference counting or other operations we would normally associate with storing a value to memory.

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: I think the "as" label makes sense when working with a raw pointer.

I like this:

func initialize<T>(as: T.Type, at: Int = 0, count: Int = 1, to value: T) -> UnsafeMutablePointer<T>

address.initialize(as: String.self, to: "hello")
address.initialize(as: String.self, at: 3, to: "hello")
address.initialize(as: String.self, count: 3, to: "hello")
address.initialize(as: String.self, at: 3, count: 3, to: "hello")

File: proposals/0107-unsaferawpointer.md (hunk 2 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: Yes. Let's dump forward/backward APIs.

File: proposals/0107-unsaferawpointer.md (hunk 3 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: address.load(String.self) is readable, but the "as" label is a nice cue that this operation reads from raw memory. Memory could in theory hold a different type (that happens to by layout compatible with String :)

File: proposals/0107-unsaferawpointer.md (hunk 4 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: If we use the "as" label everywhere, which I like for raw pointers, then this is more readable:

address.load(fromByteOffset: offset, as: Int.self)

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: on Wed Jul 06 2016, Andrew Trick <notifications-AT-github.com> wrote:

> I like this:
>
> func initialize<T>(as: T.Type, at: Int = 0, count: Int = 1, to value: T) -> UnsafeMutablePointer<T>
>
> address.initialize(as: String.self, to: "hello")

Changing the order would be more idiomatic, since we already have 

         "hello" as String

  address.initialize(to: "hello", as: String.self)

> 
> address.initialize(as: String.self, at: 3, to: "hello")

That's non-grammatical, and I still don't know whether the `at:`
parameter is measured in bytes or in `MemoryLayout<String>.stride`s.
Which is it?  I can't even make a good recommendation until I know your
intentention.

> address.initialize(as: String.self, count: 3, to: "hello")

I would prefer `count:` to go last and be defaulted to 1

> address.initialize(as: String.self, at: 3, count: 3, to: "hello")

Are all these combinations really needed?  I'm just not sure it's worth
the API complexity.

Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: One other point: saying `initializeMemory` would probably be clearer and more consistent.  We are not initializing the pointer.

File: proposals/0107-unsaferawpointer.md (hunk 3 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Okay, I get your point… but if you really want to make that clear, you might want to go with

```swift
address.load(interpretingMemoryAs: String.self)
```

File: proposals/0107-unsaferawpointer.md (hunk 4 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Agreed.

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: PING!

File: proposals/0107-unsaferawpointer.md (hunk 6 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: That’s already a precondition of the function.  If anything I would encode *that* in the name, e.g. `store(x, asTrivial: Int.self)`

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: I'll try to always include a link to the [full doc comments](https://github.com/atrick/swift/blob/80a1f2624ee4aae7d51253d1b2541826b01d4118/stdlib/public/core/UnsafeRawPointer.swift.gyb). This PR probably should have been done off the branch rather than the proposal itself.

  /// Load a single value of type `T` at the specified `index` from
  /// this memory as if it contains at least `index` + 1 contiguous
  /// values of type `T`.

The `load(fromContiguous:at:) API is a convenience to avoid typing fromByteOffset: strideof(T.self) * index.

The original request was for a load(fromArrayOf: T.Type, at: Int) API. But I don't want to use the term "array" so I've been using "contiguous"

File: proposals/0107-unsaferawpointer.md (hunk 6 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: Brent is correct.

TL;DR: Storing into raw memory is significantly more raw than loading from raw memory. `storeRaw` is basically our exposed `memcpy` API.

From the proposal:

The `load` and `storeRaw` operations are assymetric. `load` reads raw
bytes but properly constructs a new value of type `T` with its own
lifetime. Any copied references will be retained. In contrast,
`storeRaw` only operates on a value's raw bytes, writing them into
untyped memory. The in-memory copy will not be constructed and any
previously initialized value in this memory will not be deinitialized
(it cannot be because its type is unknown). Consequently, `storeRaw`
should only be performed on trivial types.

Assigning memory to a nontrivial type is done by binding the type:

```swift
rawPtr.bindMemory(to: PreviousType.self, capacity: 1).deinitialize(count: 1)
rawPtr.initialize(NewType.self, with: NewType())
```

File: proposals/0107-unsaferawpointer.md (hunk 7 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: This is a list of methods that only changed with respect to their naming conventions. Next time I'll send a PR on the branch that defines UnsafeRawPointer for more thorough review.

File: proposals/0107-unsaferawpointer.md (hunk 6 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: ...and I think Dave's point is that if the method is only valid on trivial types, the user doesn't really need to worry about all that.

I'm fine with `store(x, asTrivial: Int.self)`

I just don't want anyone to think that `load` & `store` are symmetric.

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: I’m not convinced this API is a good idea.  Is there really a strong enough use-case for it?  

My theory: if you are dealing with actual contiguous `T`’s you should be using typed access.  Otherwise, your `T` is unlikely to line up on a multiple of `MemoryLayout<T>.stride` anyhow.  What’s left over?  Are the uses important enough to complicate the API for?

Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: I'm working on a new PR, but meanwhile here's an update:

I'll remove load(fromContiguous:at:) and storeRaw(_:toContiguous:at:) because they read poorly, they are additive, and we don't have compelling use cases yet.

Now, for single-value reads and writes we have:

func load<T>(as: T.Type) -> T
func load<T>(fromByteOffset: Int, as: T.Type) -> T

func store<T>(_: T, asTrivial: T.Type)
func store<T>(_: T, toByteOffset: Int, asTrivial: T.Type)

It's necessary to provide these APIs because they are the only way to
access the bytes of untyped memory. They don't really satisfy common
anticipated use cases. I think the "ByteOffset" convenience variants
are reasonable if though they are additive because they make it even
more clear that the function is accessing raw bytes. For manual layout, I
would slightly prefer that people use the "ByteOffset" APIs than bytewise pointer arithmetic.

Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: Sorry for copy-pasting quickly. This is the proposed API for single-value reads and writes of raw bytes:

```swift
func load<T>(as: T.Type) -> T
func load<T>(fromByteOffset: Int, as: T.Type) -> T

func store<T>(_: T, asTrivial: T.Type)
func store<T>(_: T, toByteOffset: Int, asTrivial: T.Type)
```

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: I would like to distinguish initialization via raw pointer from initialization via typed pointer primarily because raw pointer initialization binds memory to a type.

Option 1:
--------

```swift
// n copies of a single value
rawPtr.initializeMemory(as: Int.self, count: n, to: x)
// read n copies out of memory at `ptrToInt`
rawPtr.initializeMemory(as: Int.self, from: ptrToInt, count: n)
```

Option 2:
--------

```swift
// n copies of a single value
rawPtr.initializeAndBind(as: Int.self, count: n, to: x)
// read n copies out of memory at `ptrToInt`
rawPtr.initializeAndBind(as: Int.self, from: ptrToInt, count: n)
```

File: proposals/0107-unsaferawpointer.md [hunk 8]
```diff
@@ -1477,34 +1477,36 @@ struct UnsafeMutableRawPointer : Strideable, Hashable, _Pointer {
...
-  func storeRaw<T>(contiguous: T.Type, from: UnsafeRawPointer, count: Int)
-  func storeRawBackward<T>(
-    contiguous: T.Type, from: UnsafeRawPointer, count: Int)
+  func storeRaw<T>(_: T, as: T.Type)
+  func storeRaw<T>(_: T, toContiguous: T.Type, at: Int)
+  func storeRaw<T>(_: T, toByteOffset: Int)
+  func storeRaw<T>(
+    contiguous: T.Type, from: UnsafeRawPointer, forwardToCount: Int)
+  func storeRaw<T>(
+    contiguous: T.Type, from: UnsafeRawPointer, backwardToCount: Int)
```
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: I realized I have not gotten enough feedback on this method:
 `fun storeRaw<T>(contiguous: T.Type, from: UnsafeRawPointer,
forwardToCount: Int)`

This is the `memcpy` API. It is very similar to bulk initialization:

`ptrToInt = rawPtr.initializeMemory(as: Int.self, from: srcAddress, count: n)`

But in this case, the memory does not need to be uninitialized, no
values are constructed, the memory is not bound, and no typed pointer
is returned.

For single value stores we have:
`rawPtr.store(x, asTrivial: Int.self)`

For bulk stores I think we should simply have:

`rawPtr.store(bytes: strideof(Int.self) * n, from: srcAddress)`

Proposed:
`func store<T>(bytes: Int, from: UnsafeRawPointer)`

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: on Thu Jul 07 2016, Andrew Trick <notifications-AT-github.com> wrote:

> For manual layout, I would slightly prefer that people use the
> "ByteOffset" APIs than bytewise pointer arithmetic.

  let x = p.load(fromByteOffset: 20, as: Int.self)

  let x = (p + 20).load(as: Int.self)

Just looking at these, I don't find the first one to be any clearer than
the second.  Do you?

-- 
Dave

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: on Thu Jul 07 2016, Andrew Trick <notifications-AT-github.com> wrote:

>>      -> UnsafeMutablePointer<T>
>> -  func initialize<T>(toContiguous: T.Type, atIndex: Int, with: T)
>> +  func initialize<T>(contiguous: T.Type, at: Int, to: T)
>
> I would like to distinguish initialization via raw pointer from
> initialization via typed pointer primarily because raw pointer
> initialization binds memory to a type.

They can both be viewed as doing that, can't they, since binding to a
type is idempotent?

>
> Option 1:
> --------
>
> ```swift
> // n copies of a single value
> rawPtr.initializeMemory(as: Int.self, count: n, to: x)
> // read n copies out of memory at `ptrToInt`
> rawPtr.initializeMemory(as: Int.self, from: ptrToInt, count: n)
> ```
>
> Option 2:
> --------
>
> ```swift
> // n copies of a single value
> rawPtr.initializeAndBind(as: Int.self, count: n, to: x)
> // read n copies out of memory at `ptrToInt`
> rawPtr.initializeAndBind(as: Int.self, from: ptrToInt, count: n)
> ```

Option 1 is much better, IM (subjective) O.

-- 
Dave

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: @jckarter requested the load-from-byte-offset feature as a significant improvement over `(p + offset).load(as:)`. I think Joe's main argument is that if `offset` is a complex expression, the `fromByteOffset` form will be much more readable. I agree.

I'm not particularly interested in very seldom used, purely additive APIs, but in this case I do think it the `byteOffset` form, for both load and store, makes it more clear that offset is in bytes and that the address computation is untyped. Otherwise, the user needs to infer that all from the pointer type.

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (atrick) [inline proposals/0107-unsaferawpointer.md]: Yes, binding is idempotent!
There is an important difference between `rawPtr.initializeMemory(as: T.self, to: T())` and `ptrToT.initialize(to: Pointee())` because the raw initialize can be called to `rebind` to a different type. The later has a precondition that the memory is already bound to `Pointee`.

That difference is somewhat signaled by the `as` label. However, I think it's also good to give the `raw` version a different name to clearly distinguish them. It then reads nicely as "initialize memory as `T`" because changing the memory's bound type is as important to this operation as changing the initialized state.

I do *not* want to rename all of the `UnsafePointer<T>` APIs with the `Memory` suffix, mainly because it would be excessive, cumbersome, and wouldn't solve any real usability problem:
- initialize
- moveInitialize
- assign
- moveAssign

I think "memory" is the obviously implied target of this API family.

File: proposals/0107-unsaferawpointer.md (hunk 5 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: on Fri Jul 08 2016, Andrew Trick <notifications-AT-github.com> wrote:

>>      -> UnsafeMutablePointer<T>
>>  
>> -  func load<T>(_: T.Type) -> T
>> -  func load<T>(_: T.Type, atByteOffset: Int) -> T
>> -  func load<T>(fromContiguous: T.Type, atIndex: Int) -> T
>> +  func load<T>(as: T.Type) -> T
>> +  func load<T>(fromByteOffset: Int, as: T.Type) -> T
>> +  func load<T>(fromContiguous: T.Type, at: Int) -> T
>
> @jckarter requested the load-from-byte-offset feature as a significant
> improvement over `(p + offset).load(as:)`. I think Joe's main argument
> is that if `offset` is a complex expression, the `fromByteOffset` form
> will be much more readable. I agree.

We have normal ways of dealing with that situation:

    let byteOffset = a complex expression
    (p + byteOffset).load(as: Int.self)

There's no need to design special provisions into this API for doing
what every Swift programmer knows how to do.

> I'm not particularly interested in very seldom used, purely additive
> APIs, but in this case I do think it the `byteOffset` form, for both
> load and store, makes it more clear that offset is in bytes and that
> the address computation is untyped. Otherwise, the user needs to infer
> that all from the pointer type.

Then are we planning to omit arithmetic on UnsafeRawPointer altogether?
If not, why not?

File: proposals/0107-unsaferawpointer.md (hunk 1 above)
Comment (dabrahams) [inline proposals/0107-unsaferawpointer.md]: Memory is the target of the initialize family but not the others.  The others are operating on initialized `T`s.

Once you initialize memory to a particular type, it seems thoroughly obvious that the memory is bound to that type.  If any word is dispensable here, I’d say it’s “bind”.
